### ISSUE TITLE
CPU softmax performance poor when dim is not the last dimension

### ISSUE THREAD

**Issue by jamesr66a** [pytorch/pytorch#34728]
```
import torch
import torch.nn.functional as F
import time

x = torch.rand(1, 3072, 1024)

NITER = 100

s = time.time()
for _ in range(NITER):
    out = F.softmax(x, dim=1)
e = time.time()

time_per_iter = (e - s) / NITER
print('time per iter (ms)', time_per_iter / 1000)
print('GB/s', (x.numel() * x.element_size() + out.numel() * out.element_size()) / time_per_iter / 1e9)
```

Gives result
```
$ OMP_NUM_THREADS=1 perf record $(which python) softmax_bench.py
time per iter (ms) 0.00013983171939849855
GB/s 0.17997221308765673
```

which is way below peak machine bandwidth.

On machine https://gist.github.com/jamesr66a/f05b2cfd879e205032d4ab7637290d40

Snooping around in perf shows we're spending the most time in scalar transcendental operations:

![image](https://user-images.githubusercontent.com/4685384/76652383-bac5b000-6523-11ea-94e5-fe4af1ce7650.png)

Conceivably this is an "outer reduction" for the purpose of summing up the exp()'ed values. We could do a vectorized reduction keeping those values in vector registers and then apply them elementwise to each slice of the tensor
